pkg/reconciler/v1alpha1/eventlistener/eventlistener_test.go

```diff
@@ -451,20 +454,20 @@ func Test_reconcileDeployment(t *testing.T) {
 			name: "deployment-replica-update",
 			startResources: test.Resources{
 				Namespaces:     []*corev1.Namespace{namespaceResource},
-				EventListeners: []*v1alpha1.EventListener{eventListener1},
-				Deployments:    []*appsv1.Deployment{deployment3},
+				EventListeners: []*v1alpha1.EventListener{eventListener3},
```

Comment:
One issue here is that before we had `spec.replicas` in the EL itself, users had to manually update the replica on the deployment itself. And we had code and this test to verify that we do not overwrite any changes to the number of replicas on an EL. So, this is a change in our behavior.  I am thinking we can either:

1. Honor the old replica changes i.e. if someone changes the existing replica numbers in the deployment itself and `spec.Replicas` is not set on the EL, we should not set it back. This keeps the behavior backwards compatible

2. Or we deprecate the existing behavior (not sure how many people depend on this behavior) and only honor EL's `spec.replicas` -- we just call this out in the release notes. 


One thing I'm curious about (and don't know) is how this will work with Horizontal Pod Autoscalers. Do they change the deployment's spec.Replicas? In that case, if an HPA scales the deployment we do not want to inadvertently scale it back down.

Reply:
Can we have the behavior where if el.spec.Replicas is set, then we track changes only in el.spec.Replicas and don't care what user set in deployment.
If el.spec.Replicas isn't set, then we don't change deployment replica and let k8s controller take care of it?
HPA uses deployment's number as you surmised. So design 2 will break them.

Reply:
Ya this is something need to verify with HPA 
Thanks for pointing it out i will try and update here

Reply:
> One thing I'm curious about (and don't know) is how this will work with Horizontal Pod Autoscalers. Do they change the deployment's spec.Replicas? In that case, if an HPA scales the deployment we do not want to inadvertently scale it back down.

@dibyom Yes HPA uses deployment so with this PR change if HPA increases replicas on deployment whenever condition met But that will be again overridden by `el.Spec.Replicas` if set as part of eventlistener yaml

Reply:
How about below opinion on supporting replicas

1. Lets keep `spec.replicas` which can be set by the user and we will add a note in README that if user sets `spec.replicas` as part of eventlistener yaml then it will override replicas values if user manually set by `kubectl edit deployment` or by setting HPA  
Which means if `spec.replicas`  set by event listener yaml then it won't respects replicas values edited by user manually or through any other mechanism (ex: HPA)

2. If user Don't set `spec.replicas` in eventlistener yaml then behavior will be same like earlier how it use to work before adding replicas field to eventlistener yaml

Few points from WG discussion

1. Don't allow `spec.replicas` as part of eventlistener yaml instead hardcode replicas to some number instead of `1` 
    -> With this approach if we hardcode `replicas` to say `3` then if HPA configured with `--min=2 --max=5` then again replicas number varies 

2. Some prior art / similar behavior for k8s addons - https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/addon-manager
    -> This is bit tricky to implement 

/cc @dibyom @khrm @wlynch

Reply:
Thanks for summarizing the discussion. It sounds good to me!